src/GitHub.App/Factories/UIFactory.cs

```diff
@@ -0,0 +1,103 @@
+﻿using GitHub.Exports;
+using GitHub.Models;
+using GitHub.UI;
+using GitHub.ViewModels;
+using NullGuard;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Reactive.Disposables;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GitHub.App.Factories
+{
+    [Export(typeof(IUIFactory))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class UIFactory : IUIFactory
+    {
+        readonly IExportFactoryProvider factory;
+
+        [ImportingConstructor]
+        public UIFactory(IExportFactoryProvider factory)
+        {
+            this.factory = factory;
+        }
+
+        /// <summary>
+        /// Creates View/ViewModel instances for the specified <paramref name="viewType"/> if they
+        /// haven't been created yet in the current flow
+        /// </summary>
+        /// <param name="viewType"></param>
+        /// <returns>true if the View/ViewModel didn't exist and had to be created</returns>
+        public IUIPair CreateViewAndViewModel(UIViewType viewType)
+        {
+            return new UIPair(viewType, factory.GetView(viewType), factory.GetViewModel(viewType));
+        }
+    }
+
+    /// <summary>
+    /// This class holds ExportLifetimeContexts (i.e., Lazy Disposable containers) for IView and IViewModel objects
+    /// A view type (login, clone, etc) is composed of a pair of view and viewmodel, which this class represents.
+    /// </summary>
+    public class UIPair : IUIPair
+    {
+        ExportLifetimeContext<IView> view;
+        ExportLifetimeContext<IViewModel> viewModel;
+        CompositeDisposable handlers = new CompositeDisposable();
+        UIViewType viewType;
```

Comment:
Whoops, I see that you `view` and `viewModel` to null in the `Dispose` method. That's probably not necessary though, is it?